Barotrauma/BarotraumaClient/ClientSource/Screens/SinglePlayerMissionSetupUI.cs

```diff
@@ -0,0 +1,319 @@
+            rightColumn.Recalculate();
+
+            pageContainer.RecalculateChildren();
+            pageContainer.GetAllChildren().ForEach(c => c.ClampMouseRectToParent = true);
+            pageContainer.GetAllChildren<GUIDropDown>().ForEach(dd =>
+            {
+                dd.ListBox.ClampMouseRectToParent = false;
+                dd.ListBox.Content.ClampMouseRectToParent = false;
+            });
+        }
+
+        public void RandomizeSeed() => seedBox.Text = ToolBox.RandomSeed(8);
+
+        public void UpdateSubList(IEnumerable<SubmarineInfo> submarines)
+        {
+            List<SubmarineInfo> subsToShow;
+            if (subFilter != CategoryFilter.All)
+            {
+                subsToShow = submarines.Where(s => s.IsCampaignCompatibleIgnoreClass && s.IsVanillaSubmarine() == (subFilter == CategoryFilter.Vanilla)).ToList();
+            }
+            else
+            {
+                string downloadFolder = Path.GetFullPath(SaveUtil.SubmarineDownloadFolder);
+                subsToShow = submarines.Where(s => s.IsCampaignCompatibleIgnoreClass && Path.GetDirectoryName(Path.GetFullPath(s.FilePath)) != downloadFolder).ToList();
+            }
+
+            subsToShow.Sort((sub1, sub2) => sub1.SubmarineClass.CompareTo(sub2.SubmarineClass) * 100 + sub1.Name.CompareTo(sub2.Name));
+
+            subList.ClearChildren();
+
+            foreach (SubmarineInfo sub in subsToShow)
+            {
+                GUITextBlock textBlock = new(new(new Vector2(1, 0.15f), subList.Content.RectTransform, minSize: new(0, 30)), ToolBox.LimitString(sub.DisplayName.Value, GUIStyle.Font, subList.Rect.Width - 65), style: "ListBoxElement")
+                {
+                    ToolTip = sub.Description,
+                    UserData = sub
+                };
+
+                if (!sub.RequiredContentPackagesInstalled)
+                {
+                    textBlock.TextColor = Color.Lerp(textBlock.TextColor, Color.DarkRed, 0.5f);
+                    textBlock.ToolTip = TextManager.Get("ContentPackageMismatch") + "\n\n" + textBlock.ToolTip.SanitizedString;
+                }
+
+                new GUITextBlock(new(new Vector2(0.5f, 0.5f), textBlock.RectTransform, Anchor.CenterRight), TextManager.Get($"submarineclass.{sub.SubmarineClass}"), font: GUIStyle.SmallFont, textAlignment: Alignment.CenterRight)
+                {
+                    TextColor = textBlock.TextColor * 0.8f,
+                    ToolTip = textBlock.ToolTip
+                };
+#if !DEBUG
+                if (!GameMain.DebugDraw && !sub.IsCampaignCompatible)
+                {
+                    textBlock.CanBeFocused = false;
+                    textBlock.TextColor *= 0.5f;
+                }
+#endif
+            }
+
+            if (subsToShow.Where(s => s.IsCampaignCompatible).ToList() is { Count: > 0 } validSubs)
+            {
+                subList.Select(validSubs.GetRandomUnsynced());
+            }
+        }
+    }
+}
```

Comment:
There is a huge amount of code duplication between this class and CampaignSetupUI. The shared logic should most likely be moved to a shared base class.